pkg/querier/querier.go

```diff
@@ -133,52 +126,106 @@ func (q *Querier) SelectSamples(ctx context.Context, params logql.SelectSamplePa
-		if err != nil {
-			return nil, err
-		}
-	case q.cfg.IngesterQueryStoreMaxLookback > 0:
-		adjustedEnd := params.End.Add(-q.cfg.IngesterQueryStoreMaxLookback)
-		if params.Start.After(adjustedEnd) {
-			chunkStoreIter = iter.NoopIterator
-			break
-		}
+	iters := []iter.SampleIterator{}
+	if ingesterQueryInterval != nil {
 		// Make a copy of the request before modifying
-		// because the initial request is used below to query ingesters
+		// because the initial request is used below to query stores
 		queryRequestCopy := *params.SampleQueryRequest
 		newParams := logql.SelectSampleParams{
 			SampleQueryRequest: &queryRequestCopy,
 		}
-		newParams.End = adjustedEnd
-		chunkStoreIter, err = q.store.SelectSamples(ctx, newParams)
+		newParams.Start = ingesterQueryInterval.start
+		newParams.End = ingesterQueryInterval.end
+
+		ingesterIters, err := q.ingesterQuerier.SelectSample(ctx, newParams)
 		if err != nil {
 			return nil, err
 		}
-	default:
-		chunkStoreIter = iter.NoopIterator
 
-	}
-	// skip ingester queries only when QueryIngestersWithin is enabled (not the zero value) and
-	// the end of the query is earlier than the lookback
-	if !shouldQueryIngester(q.cfg, params) {
-		return chunkStoreIter, nil
+		iters = append(iters, ingesterIters...)
 	}
 
-	iters, err := q.ingesterQuerier.SelectSample(ctx, params)
-	if err != nil {
-		return nil, err
+	if storeQueryInterval != nil {
+		params.Start = storeQueryInterval.start
+		params.End = storeQueryInterval.end
+
+		storeIter, err := q.store.SelectSamples(ctx, params)
+		if err != nil {
+			return nil, err
+		}
+
+		iters = append(iters, storeIter)
 	}
-	return iter.NewHeapSampleIterator(ctx, append(iters, chunkStoreIter)), nil
+	return iter.NewHeapSampleIterator(ctx, iters), nil
 }
 
-func shouldQueryIngester(cfg Config, params logql.QueryParams) bool {
-	lookback := time.Now().Add(-cfg.QueryIngestersWithin)
-	return !(cfg.QueryIngestersWithin != 0 && params.GetEnd().Before(lookback))
+func (q *Querier) buildQueryIntervals(queryStart, queryEnd time.Time) (*interval, *interval) {
+	// limitQueryInterval is a flag for whether store queries should be limited to start time of ingester queries.
+	limitQueryInterval := false
+	// ingesterMLB having -1 means query ingester for whole duration.
+	ingesterMLB := time.Duration(-1)
+	if q.cfg.IngesterQueryStoreMaxLookback != 0 {
+		// IngesterQueryStoreMaxLookback takes the precedence over QueryIngestersWithin while also limiting the store query range.
+		limitQueryInterval = true
+		ingesterMLB = q.cfg.IngesterQueryStoreMaxLookback
+	} else if q.cfg.QueryIngestersWithin != 0 {
+		ingesterMLB = q.cfg.QueryIngestersWithin
+	}
+
+	// query ingester for whole duration.
+	if ingesterMLB == -1 {
+		i := &interval{
+			start: queryStart,
+			end:   queryEnd,
+		}
+
+		if limitQueryInterval {
+			// query only ingesters.
+			return i, nil
+		}
+
+		// query both stores and ingesters without limiting the query interval.
+		return i, i
+	}
```

Comment:
Why query ingester for whole duration ？

Reply:
![image](https://user-images.githubusercontent.com/16623135/144800906-38b03d2e-e049-4310-9e13-a989c96ed1a3.png)

The logic here is inconsistent with the documentation.

Reply:
This is to support the rare usecase where you run Loki with multiple single binaries and each of them storing the data on local filesystem without any sharing.